analysis/task_analysis_surface/second_level_analysis_surf_ss.py

```diff
@@ -0,0 +1,149 @@
+        fx_results = _compute_fixed_effects_params(
+            np.squeeze(
+                [nb.load(fname).get_fdata(dtype='f4') for fname in effect_size_list]
+            ),
+            np.squeeze(
+                [nb.load(fname).get_fdata(dtype='f4') for fname in effect_variance_list]
+            ),
+            precision_weighted=False)
+        
+        output_dict[contrast]=fx_results 
+    
+    return output_dict
+
+
+
+def save_second_level_outputs(sub,task,ses):
+    print(f'Now processing subject {sub} for task {task} in session {ses}')
+    output_dict = run_second_level_model(sub,task,ses)
+    
+    
+    #check that output does not exist
+    output = glob.glob(f'../../../derivatives/task_analysis_surface/second_level/sub-{sub}/ses-{ses}/task-{task}/sub-{sub}_ses-{ses}_task-{task}_rec-unco_contrast-*_effect_size_fx.dscalar.nii')
+    if output: #checks if list is not empty
+        print(f'At least partial 2nd level output exists for subject {sub}, session {ses} and task {task}. This must be deleted for before generating new output.')
+        return 
+    
+    orig_dscalar_list = glob.glob(f'../../../derivatives/task_analysis_surface/first_level/sub-*/ses-{ses}/task-{task}/sub-*_ses-{ses}_task-{task}_rec-unco_run-*_contrast-*_effect_size.dscalar.nii')
+    orig_orig_dscalar_nimg = nimg.load_img(orig_dscalar_list[0])
+    time_axis, brain_model_axis = [orig_orig_dscalar_nimg.header.get_axis(i) for i in range(orig_orig_dscalar_nimg.ndim)]
+
+    for contrast in output_dict.keys():
+        #calculate contrast maps for all contrasts
+        contrast_output = output_dict[contrast]
+        dscalar_output_dict = create_dscalar(brain_model_axis, contrast_output) 
+        
+        #create paths to output dir if not exist
+        derivatives_path = '../../../derivatives'
+        nilearn_output_path = os.path.join(derivatives_path, 'task_analysis_surface','second_level',f'sub-{sub}',f'ses-{ses}',f'task-{task}')
+        if not os.path.isdir(nilearn_output_path):
+            os.makedirs (nilearn_output_path)
+            
+        #save contrast maps to files
+        dscalar_output_dict['effect_size_fx'].to_filename(f'../../../derivatives/task_analysis_surface/second_level/sub-{sub}/ses-{ses}/task-{task}/sub-{sub}_ses-{ses}_task-{task}_rec-unco_contrast-{contrast}_effect_size_fx.dscalar.nii')
+        dscalar_output_dict['effect_variance_fx'].to_filename(f'../../../derivatives/task_analysis_surface/second_level/sub-{sub}/ses-{ses}/task-{task}/sub-{sub}_ses-{ses}_task-{task}_rec-unco_contrast-{contrast}_effect_variance_fx.dscalar.nii')
+        dscalar_output_dict['stat_fx'].to_filename(f'../../../derivatives/task_analysis_surface/second_level/sub-{sub}/ses-{ses}/task-{task}/sub-{sub}_ses-{ses}_task-{task}_rec-unco_contrast-{contrast}_stat_fx.dscalar.nii')
+    
+    print(f'{sub} {task} {ses}: processing done')
+        
+        
+        
+def create_dscalar(brain_model_axis, contrast_output):
```

Comment:
some of these functions could be put into a single python utils library and used across the scripts. not essential for submission, but minimizes errors.

Reply:
I did not know about this option, but good to know for next time! i agree, i did not like having to copy the same function into several scripts